operator/pkg/awsprovider/instances/reconciler.go

```diff
@@ -180,24 +180,36 @@ func (c *Controller) subnetsFor(ctx context.Context, dataplane *v1alpha1.DataPla
 	if err != nil {
 		return nil, fmt.Errorf("getting subnet for %s, %w", dataplane.Spec.ClusterName, err)
 	}
-	return c.filterPrivateSubnets(ctx, subnetIDs)
+	// Returns public subnets if no private subnets are found
+	return c.filterSubnets(ctx, subnetIDs)
 }
 
-func (c *Controller) filterPrivateSubnets(ctx context.Context, ids []*string) ([]string, error) {
+func (c *Controller) filterSubnets(ctx context.Context, ids []*string) ([]string, error) {
+	if len(ids) == 0 {
+		return nil, fmt.Errorf("found zero subnets while filtering")
+	}
 	output, err := c.ec2api.DescribeSubnetsWithContext(ctx, &ec2.DescribeSubnetsInput{
 		SubnetIds: ids,
 	})
 	if err != nil {
 		return nil, fmt.Errorf("describing subnet, %w", err)
 	}
-	result := []string{}
+	public := []string{}
+	private := []string{}
 	for _, subnet := range output.Subnets {
```

Comment:
Check out https://github.com/samber/lo. 

```
public := lo.Filter(subnets, func(subnet *ec2.Subnet, _ int) { return ptr.BoolValue(subnet.MapPublicIpOnLaunch) })
private := lo.Reject(subnets, func(subnet *ec2.Subnet, _ int) { return ptr.BoolValue(subnet.MapPublicIpOnLaunch) })
if len(private) > 0 {
  return private
}
return public
```

Reply:
Is it ever the case that we might want to mix and match public and private subnets?

Reply:
If private subnets are available they are preferred.

Reply:
@prateekgogia do we want to take the dependency on `samber/lo`?

Reply:
Usually, I take a dependency on other modules if we have logic thats being used at more than 1 place / can be error prone. 
In this particular case, I think its straightforward and easy to understand what you are trying to do. IMO, we can keep at it as-is.